Fix LWT insert crash if clustering key is null

gusev-p opened the pull request:
[PR](https://github.com/scylladb/scylladb/pull/9314) fixed a similar issue with regular insert statements
but missed the LWT code path.

It's expected behaviour of
`modification_statement::create_clustering_ranges` to return an
empty range in this case, since `possible_lhs_values` it
uses explicitly returns `empty_value_set` if it evaluates `rhs`
to null, and it has a comment about it (All NULL
comparisons fail; no column values match.) On the other hand,
all components of the primary key are required to be set,
this is checked at the prepare phase, in
`modification_statement::process_where_clause`. So the only
problem was `modification_statement::execute_with_condition`
was not expecting an empty `clustering_range` in case of
a null clustering key.

Also this patch contains a fix for the problem with wrong
column name in Scylla error messages. If `INSERT` or `DELETE`
statement is missing a non-last element of
the primary key, the error message generated contains
an invalid column name.

The problem occurs if the query contains a column with the list type,
otherwise
`statement_restrictions::process_clustering_columns_restrictions`
checks that all the components of the key are specified.

nyh commented on test/cql-pytest/test_query_validation.py:
```diff
@@ -0,0 +1,41 @@
+def test_insert_with_list_column_and_missing_clustering_key_part(cql, table1):
+    key = unique_key_string()
+    with pytest.raises(InvalidRequest, match='Missing mandatory PRIMARY KEY part a'):
+        cql.execute(cql.prepare(f"INSERT INTO {table1} (id,b,c) VALUES ('{key}',1,null)"))
```
When I read the commit message, I thought to myself that I didn't understand why it is specific to collection columns. But now, reading this test, I had a thought:
With non-frozen collections, an **empty list** is identical to a null (a missing list). So maybe a null list was intentionally allowed - to support the ability to have an empty list? Or, maybe the other way around - and empty list should not be allowed? I have to admit I'm not sure. I'm not sure I even knew that it was legal to have a non-frozen collection as part of a key! In any case, the first thing I would check is what Cassandra does in this case - and this is why we have "run-cassandra" that I asked you above to run.
In any case, if the purpose of this test is to check this meaning of a null collection as part of a key, it would be good to explain this in a comment above the test.

nyh replied:
Also, maybe in the same test you should also try a null for "b", which should also fail with an error? Even if we didn't have a bug in this case (and only had a bug in the case of "c"), we should have a regression test to check such a bug doesn't suddenly happen in the future.

nyh commented on test/cql-pytest/test_query_validation.py:
```diff
@@ -0,0 +1,41 @@
+    with pytest.raises(InvalidRequest, match="Primary key column 'a' must be specified in order to modify column 'c'"):
+        cql.execute(cql.prepare(f"DELETE c FROM {table1} WHERE id='{key}' AND b=1"))
+    cql.execute(cql.prepare(f"DELETE c FROM {table1} WHERE id='{key}' AND a=1 AND b=1"))
+    assert list(cql.execute(f"SELECT c FROM {table1} WHERE id='{key}' AND a=1 AND b=1"))[0][0] is None
```
I'm surprised this `...[0][0] is None` is even allowed for an empty list (I assumed Python would throw on `[0]` to an empty list). I think it will look clearer if you checked that the list is empty, e.g., `list(...) == []`.

nyh commented on test/cql-pytest/test_query_validation.py:
```diff
@@ -0,0 +1,41 @@
+def test_insert_with_list_column_and_missing_partition_key_part(cql, table2):
+    key = unique_key_string()
+    with pytest.raises(InvalidRequest, match='Missing mandatory PRIMARY KEY part id1'):
+        cql.execute(cql.prepare(f"INSERT INTO {table2} (id2,a,c) VALUES ('{key}',1,[1])"))
```
Also here, since you already suspected we have problems with maybe missing the first component is different from missing the second component, that you also add in this test a check that an INSERT with the first component missing also results in the correct error.

nyh reviewed (changes requested):
Thanks, very good.
I just have relatively minor requests on the second patch.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3100/

kostja reviewed (approved):
Agree with Nadav's comments that the test coverage could be extended and explained. I think we can reuse an existing test file. unrestricted_column() should have a comment warning/check that it is used at prepare step only, since it's expensive. 
Otherwise LGTM, thanks for fixing both issues - LWT and improper message.

nyh commented:
>  I think we can reuse an existing test file. 

I agree with that, and thought of commenting on this myself, but at the end didn't want to be too pedantic so I left it out of my comments :-) 

In this case, we already have a test file `test/cql-pytest/test_null.py` which has a lot of tests related to null values - including tests like `test_insert_missing_key`, `test_insert_null_key`, `test_insert_null_key_lwt`, `test_delete_null_key`. Don't you agree that your new tests fit in that file well?

By the way, the problem with a new test file is not that it costs us something - unlike the C++ tests that have huge per-file compilation-time and executable-size overheads, in Python tests there is no extra cost for having more test files. Rather, the problem of introducing new test files is it becomes very unclear what is the purpose of each individual file, and where developers should put new tests or where they should look for tests. To use your example, which tests in the future should belong in the new file called "test_query_validation"? After all, most tests of most CQL features also include some tests for validation of incorrect uses. If you do create a new test file, please try to follow my lead of starting the test file with a comment explaining its purpose. If you find that hard to do, maybe that file shouldn't exist :-)

Oh - and the test file should also begin with a copyright blurb.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3105/

nyh commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3105/

Same compilation failure as before. Please try to build yourself with "dbuild" in case the code builds correctly on your build machine but not in Jenkins.

gusev-p commented:
v2:
  * compilation fix, new fmt lib [prohibits implicit scoped enums conversons](https://github.com/fmtlib/fmt/commit/fd62fba9855c9cf8fe17eb9363e03f93563d9dc3)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3112/

gusev-p commented:
Investigated the error message issue a little bit further, seems there are two other problems:
  * the error message doesn't include the column name, Scylla message: `Invalid null value in condition for a key column`, cassandra message `Invalid null value in condition for column p`;
  * `statement_restrictions::range_or_slice_eq_null` ignores `IN` operator, so
```
q = cql.prepare(f"DELETE FROM {table1} WHERE p in ?")
with pytest.raises(InvalidRequest, match='null value'):
    cql.execute(q, [(None)])
```
doesn't generate any errors on Scylla, but produces the same `Invalid null value in condition for column p` on cassandra.

I tried to fix them, but got stuck over `partially_deserialize_listlike`, it doesn't seem to be able to deserialize `null` inside collections. `read_collection_value` contains an explicit `throw("null is not supported inside collections")`, but when I pass None from the driver is is silently converted to empty byte string somewhere. I'm not sure if it makes sense to investigate further. @nyh @kostja

gusev-p commented:
And there is a third problem with batches, Scylla ignores nulls in conditions, but Cassandra generates the same `Invalid null value in condition for column p`.

nyh commented:
> Investigated the error message issue a little bit further, seems there are two other problems:
> 
>     * the error message doesn't include the column name, Scylla message: `Invalid null value in condition for a key column`, cassandra message `Invalid null value in condition for column p`;

It's not strictly necessary that our error messages be 100% identical to Cassandra, and although adding the column name to the error message would be an improvement (so if you can do it, I won't complain), we won't call it a "bug fix". Until then, when we write a test if we check the error message we want to only check the identical part (in this case "Invalid null value") and not the whole message, which is different in Scylla and Cassandra.

>     * `statement_restrictions::range_or_slice_eq_null` ignores `IN` operator, so 
> 
> ```
> q = cql.prepare(f"DELETE FROM {table1} WHERE p in ?")
> with pytest.raises(InvalidRequest, match='null value'):
>     cql.execute(q, [(None)])

What do the extra parentheses do here? If I understand correctly, this is a tuple containing a null, not a list containing a null. 

> ```
> 
> doesn't generate any errors on Scylla, but produces the same `Invalid null value in condition for column p` on cassandra.
> 
> I tried to fix them, but got stuck over `partially_deserialize_listlike`, it doesn't seem to be able to deserialize `null` inside collections. `read_collection_value` contains an explicit `throw("null is not supported inside collections")`, but when I pass None from the driver is is silently converted to empty byte string somewhere. I'm not sure if it makes sense to investigate further. @nyh @kostja

If you find bugs or incompatibilities with Cassandra that you can't or don't want to immediately fix, please consider opening a seperate issue on each such issue and if possible, create an xfailing test reproducing it.

Please think of the issue I mentioned in my review - whether these things are a general problem of nulls, nulls inside a tuple (I think this is what you had above), null _inside_ a list, or a null list (i.e., empty list). These can be separate problems, and it's important we understand the real scope of the problems we discover.

If the result will be 5 different issues with 5 different tests reproducing them, it would be good. Not good that we have 5 bugs, but good we discovered them.

nyh commented:
> If the result will be 5 different issues with 5 different tests reproducing them, it would be good. Not good that we have 5 bugs, but good we discovered them.

In any case, obviously you don't need to fix all the problems in one patch or even one PR. You can have a patch that solves just one bug, leaving 5 other bugs in place. Just one thing I would ask is that if you took the time to research this issue, and found those 5 other bugs, that if you're not planning to solve them right now - that you leave behind issues and (if you're up to it) reproducing tests so the effort that you put into this research won't go to waste.

kostja commented:
> > Investigated the error message issue a little bit further, seems there are two other problems:
> > ```
> > * the error message doesn't include the column name, Scylla message: `Invalid null value in condition for a key column`, cassandra message `Invalid null value in condition for column p`;
> > ```
> 
> It's not strictly necessary that our error messages be 100% identical to Cassandra, and although adding the column name to the error message would be an improvement (so if you can do it, I won't complain), we won't call it a "bug fix". Until then, when we write a test if we check the error message we want to only check the identical part (in this case "Invalid null value") and not the whole message, which is different in Scylla and Cassandra.

The message has to have a column name for multi-part keys, otherwise it's very hard to say which part is missing. Besides, it already contains the column name, but the name is wrong - it's referring to a wrong column. So it's a real bug.

gusev-p commented on test/cql-pytest/test_query_validation.py:
```diff
@@ -0,0 +1,41 @@
+    cql.execute("DROP TABLE " + table)
+
+
+def test_insert_with_list_column_and_missing_clustering_key_part(cql, table1):
```
> Please add a comment explaining the purpose of each test, or at least one comment for a group of tests. In particular, if a test reproduces an issue (in this case https://github.com/scylladb/scylladb/issues/12046), please say this in the comment.

Ok, but I don't get the value of it. The name of the test fully describes what's happening, and the issue/pr can be extracted from VCS. Is this a standard convention in the Scylla repo?

nyh replied:
Since I wrote about 90% of the tests in test/cql-pytest and test/alternator, I have taken certain liberties in defining the "standard conventions" there by myself :-)

But to give you a serious answer: Tests are rarely changed in their purpose. What is much more common is that at some point after a Scylla code change some test begins to fail, and a random developer - who probably never saw this test before - now needs to understand what this test did and why it failed. This developer wants to know why this test was even correct - beyond the obvious fact that we know it used to pass in the past and now doesn't. Even for trivial tests, it is often not obvious from the code why the person who wrote the test wanted to test specifically this thing. E.g., in this case the name of the function has the word "list", but why - why is list an important part of this test? Or maybe it isn't, and it's just an example? Moreover, was this test reproducing a known bug (for which we have an issue number) or just something that someone wrote "just in case"? Knowing this is super-useful when running tests against older branches - and then if we see a test failing that claims to be reproducing issue #12046, we immediately know that we forgot to backport the fix for that issue in that branch, and don't need to start guessing what this test failure might mean.

You're right that the developer can piece together some of this information from "git blame" and "git show". But not always all of it, and not easily. Especially when the test comes separately from the fix so the test commit doesn't have "Fixes" of the bug itself, so the information that this test is testing #12046 may not appear in the commit message itself. I personally consider relying on git instead of comments to be a bad practice. One of the annoyances is that if the tests are ever moved around, for example, it becomes a real chore trying to hunt down the commit message which originally introduced a test to see it's explanation - while the comment moves around with the test. 
It happened to me just today when I was looking at dtest on a failing test from several years ago - a test which got moved around (and translated from nosetest to pytest) so the original commit that introduced that test originally wasn't easy to find.

Please take a look at other files in test/cql-pytest and see some examples of comments. Take a look at a random test which you're not familiar with it, don't read the comment and see if you can guess why this test exists - and then look at the comment. I hope that this will convince you that writing comments on tests is very useful.

gusev-p replied:
Ok, thanks a lot for the detailed comment!

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3134/

gusev-p commented:
v3:
  * addressed @nyh review, added comments to the tests, added some tests
  * add column name to the error message in case of null primary key
  * addressed other bugs I've found (null validation in batches, `DELETE` with null value for IN parameter)
  * moved all the tests into `test_null.py`

@nyh could you please have another look?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3140/

nyh commented:
> Not sure how to deal with CI failure:
> 
> ```
> ERROR: Could not get unified-pack-url-x86_64 from relocatable metadata file
> 19:55:37  Finished: FAILURE
> ```
> 
> just restarted it for now.

It didn't help :-( Another error message I see:
```
[2022-11-23T20:52:02.775Z] grep: /jenkins/workspace/releng/Scylla-CI/00-Build.txt: No such file or directory
```

Obviously this isn't related to your patch and is some problem in the CI code, but I don't know what this means, maybe rebasing your patch will get you some unrelated fixes and it will work? But I don't know :-(

nyh commented on test/cql-pytest/test_null.py:
```diff
@@ -65,6 +79,51 @@ def test_insert_null_key_lwt(cql, table1):
+# If INSERT or DELETE statement is missing a non-last element of
+# the primary key, the error message generated contained
+# an invalid column name (#12046).
+# The problem occurred if the query contains a column with the list type,
```
Thanks. The tests look good, but I'm still not sure I understand why these tests need to use "c" as a list. If I understand correctly, the tests revolve around missing clustering key or partition key parts - and those are all "text" columns in your tests, not lists. Why does the type of C matter?

More on this below.

gusev-p replied:
The reason is dictated by the current structure of the code. If the statement contains list columns, the `modification_statement::_selects_a_collection` variable will be `true`, in `modification_statement::process_where_clause` it will be passed to `statement_restrictions::statement_restrictions` as `bool for_view = true` (this looks strange and seems like a bug, but nevertheless), then in `statement_restrictions::process_clustering_columns_restrictions`, since `for_view==false`, we will skip the check that all the components of the clustering key are set. So, the problem occurs only if the query contains a list column.

nyh replied:
Wow. Ok. This "for_view = true" thing looks very suspicious but I don't know if you still have the energy to continue chasing this (the list of commit grew to 6, can now continue to grow to 7 ;-) ).

gusev-p replied:
I'll look at the history of this "for_view = _selects_a_collection" thing and make another PR if I find something interesting)

nyh commented on test/cql-pytest/test_null.py:
```diff
@@ -65,6 +79,51 @@ def test_insert_null_key_lwt(cql, table1):
+    key = unique_key_string()
+    with pytest.raises(InvalidRequest,
+                       match='Missing mandatory PRIMARY KEY part a|Some clustering keys are missing: a'):
+        cql.execute(cql.prepare(f"INSERT INTO {table2} (id,b,c) VALUES ('{key}',1,null)"))
```
If I understand correctly, the important part of this test is that this INSERT is missing a value for "a".
So how does the fact that "c" - some non-key column - is assigned a null - related to this test? Is it really related? Assigning null to c, which is a list, is perfectly fine - it just assigns an empty list.

gusev-p replied:
The important part is not null assigned to list, but that the query just contains a list column. The reason is described above.

nyh commented on test/cql-pytest/test_null.py:
```diff
@@ -65,6 +79,51 @@ def test_insert_null_key_lwt(cql, table1):
+# also to not bother with providing condition for it in the WHERE clause.
+def test_delete_with_list_column_and_missing_clustering_key_part(cql, table2):
+    key = unique_key_string()
+    cql.execute(cql.prepare(f"INSERT INTO {table2} (id,a,b,c) VALUES ('{key}',1,1,[1])"))
```
Again, is it important that C is a list in this test? Or it's just an arbitrary column type and would have worked the same also for an integer?

gusev-p replied:
It's important for the reasons described above. And btw, this fact is mentioned in the comments for the `test_insert_with_list_column_and_missing_clustering_key_part` which is referenced here.

nyh reviewed (approved):
Looks good, thanks!

I left some questions where I still didn't understand about why some of the tests used a list - if it's just a relic from how you discovered the bugs or an important part of the bug.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3143/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3149/
